src/sleep/sleep.c

```diff
@@ -263,40 +263,88 @@ static int execute(
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                log_debug("Set timerfd wake alarm to estimate battery discharge rate for %s",
+                        FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+
+                timespec_store(&ts.it_value, suspend_interval);
+
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_MONOTONIC);
+
+                        current_capacity = read_battery_capacity();
+                        if (current_capacity < 0)
+                                return log_error_errno(errno, "Error fetching battery capacity: %m");
+
+                        log_debug("%d%% battery discharged in %s time",
+                                        last_capacity - current_capacity,
+                                        FORMAT_TIMESPAN(after_timestamp - before_timestamp, USEC_PER_SEC));
+                        estimated_total_discharge_time = (current_capacity * (before_timestamp - after_timestamp)) / (last_capacity - current_capacity);
+                        r = put_discharge_rate(FORMAT_TIMESPAN(estimated_total_discharge_time, USEC_PER_SEC));
+                        if (r < 0)
+                                return log_warning_errno(r, "Failed to update battery discharge rate: %m");
+
+                        log_debug("Battery will discharge in %s time", FORMAT_TIMESPAN(estimated_total_discharge_time, USEC_PER_SEC));
+                        return 0;
+                }
+                tfd = safe_close(tfd);
+
+                /* If woken up after alarm time, calculate battery capacity and battery capacity_level and estimate discharge rate */
+
+                current_capacity = read_battery_capacity();
+                if (current_capacity < 0)
+                        return log_error_errno(errno, "Error fetching battery capacity: %m");
+
+                estimated_total_discharge_time = current_capacity / (last_capacity - current_capacity);
+
+                r = put_discharge_rate(FORMAT_TIMESPAN(estimated_total_discharge_time, USEC_PER_SEC));
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to update battery discharge rate: %m");
```

Comment:
This should probably be non fatal, given we have a fallback - log but don't return (and add 'ignoring' to the log message)